woocommerce/woocommerce
# Prevent product from being saved prematurely when updated via REST API

### Changes proposed in this Pull Request:

The last step performed by `WC_REST_Products_Controller::prepare_object_for_database` when a product is about to be created or updated via REST API is to fire a `woocommerce_rest_pre_insert_product_object` filter that can return an instance of `WP_Error`, thus cancelling the product update. However if the body of the request contains information about product images `wc_product_attach_featured_image` is called, which saves the product; thus by the time  `woocommerce_rest_pre_insert_product_object` is fired it's too late to return an error, as all the product changes in the request are already persisted.

This pull request adds a new optional `$save_product` argument to `wc_product_attach_featured_image`, defaulting to true; and the call to this function in the REST API code passes the argument as false, thus preventing product changes from being persisted if the filter returns an error.

Closes #45147.



### How to test the changes in this Pull Request:

1. Create a simple product with a price of 1, take note of its id.
2. Add the following snippet at the end of `woocommerce.php` or by using [a code snippets plugin](https://wordpress.org/plugins/code-snippets/):

```php
add_filter('woocommerce_rest_pre_insert_product_object', fn() => new WP_Error("This product isn't good for some reason"), 0);
```

3. Add an image to the site gallery (in Media - Library) and take note of its id; the id appears at the end of the image permalink, if you click on "Edit image" you'll see the permalink in the address bar of your browser.
4. Create a REST API key+secret pair if you don't have one (in WooCommerce - Settings - Advanced - REST API).
5. Run the following request in command line, or the equivalent request in a tool like Postman, replacing ids as appropriate (`<API key>` is `base64($consumer_key . ':' . $consumer_secret)`), it will return an error as expected:

```
curl -X PUT -H "Content-Type: application/json" -H "Authorization: Basic <API key>" -d '{ "regular_price": "2", "images": [ { "id": <image id> }]}' http://localhost/wp-json/wc/v3/products/<product id>
```

6. Verify that there are no changes in the product: the image has not been associated to it, and the price hasn't changed.
7. Try again but passing an image URL, verify that again, there are no changes in the product:

```
curl -X PUT -H "Content-Type: application/json" -H "Authorization: Basic <API key>" -d '{ "regular_price": "2", "images": [ { "src": "https://upload.wikimedia.org/wikipedia/commons/thumb/c/cb/Canon_V-20_MSX_computer.jpg/330px-Canon_V-20_MSX_computer.jpg" }]}' http://localhost/wp-json/wc/v3/products/<product id>
```

8. Remove the snippet that hooks on `woocommerce_rest_pre_insert_product_object` and try again, verify that this time the product is updated with the new price and the image.

### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
